File: archivist/src/test/java/com/zorroa/archivist/repository/BlobDaoTests.java
package com.zorroa.archivist.repository;

import com.google.common.collect.ImmutableMap;
import com.zorroa.archivist.AbstractTest;
import com.zorroa.archivist.domain.*;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.dao.EmptyResultDataAccessException;

import java.util.List;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class BlobDaoTests extends AbstractTest {

    @Autowired
    private BlobDao blobDao;

    @Autowired
    private PermissionDao permissionDao;

    Blob blob;

    @Before
    public void init() {
        blob = blobDao.create("app", "feature", "name",
                ImmutableMap.of("foo", "bar"));
    }

    @Test
    public void testCreate() {
        assertEquals("app", blob.getApp());
        assertEquals("feature", blob.getFeature());
        assertEquals("name", blob.getName());
        assertEquals("bar", ((Map)blob.getData()).get("foo"));
    }

    @Test
    public void testUpdate() {
        assertTrue(blobDao.update(blob, ImmutableMap.of("shizzle", "mcnizzle")));
    }

    @Test(expected=DuplicateKeyException.class)
    public void testCreateDuplicate() {
        blobDao.create("app", "feature", "name",
                ImmutableMap.of("foo", "bar"));
    }

    @Test
    public void testGetAll() {
        List<Blob> all = blobDao.getAll("app", "feature");
        assertEquals(1, all.size());

        blobDao.create("app", "feature", "name2",
                ImmutableMap.of("foo", "bar"));
        all = blobDao.getAll("app", "feature");
        assertEquals(2, all.size());
    }

    @Test
    public void testGetId() {
        BlobId id = blobDao.getId("app", "feature", "name", Access.Read);
        assertEquals(id.getBlobId(), blob.getBlobId());
    }

    @Test
    public void testReplacePermissions() {
        BlobId id = blobDao.getId("app", "feature", "name", Access.Read);
        Acl acl = new Acl();
        acl.addEntry(7, 7);
        blobDao.setPermissions(id, new SetPermissions().setAcl(acl).setReplace(true));
        assertEquals(1, blobDao.getPermissions(id).size());
    }

    @Test(expected = EmptyResultDataAccessException.class)
    public void testCheckPermissions() {
        BlobId id = blobDao.getId("app", "feature", "name", Access.Write);
        Permission test = permissionDao.create(new PermissionSpec()
                .setDescription("foo")
                .setType("test")
                .setName("test"), false);

        Acl acl = new Acl();
        acl.addEntry(test, 7);
        blobDao.setPermissions(id, new SetPermissions().setAcl(acl).setReplace(true));

        authenticate("user");
        id = blobDao.getId("app", "feature", "name", Access.Write);
    }

    @Test(expected = DuplicateKeyException.class)
    public void testReplacePermissionsWithDuplicate() {
        BlobId id = blobDao.getId("app", "feature", "name", Access.Read);
        Acl acl = new Acl();
        acl.addEntry(7, 7);
        acl.addEntry(7, 7);
        blobDao.setPermissions(id, new SetPermissions().setAcl(acl).setReplace(true));
    }

    @Test
    public void testUpdatePermissions() {
        BlobId id = blobDao.getId("app", "feature", "name", Access.Read);
        Acl acl = new Acl();
        acl.addEntry(7, 7);
        blobDao.setPermissions(id, new SetPermissions().setAcl(acl).setReplace(false));
        assertEquals(1, blobDao.getPermissions(id).size());

        // add a new permission
        Permission p = permissionDao.get("group::administrator");
        acl = new Acl();
        acl.addEntry(p, 7);
        blobDao.setPermissions(id, new SetPermissions().setAcl(acl).setReplace(false));
        assertEquals(2, blobDao.getPermissions(id).size());

        // remove permission
        acl = new Acl();
        acl.addEntry(p, 0);
        blobDao.setPermissions(id, new SetPermissions().setAcl(acl).setReplace(false));
        assertEquals(1, blobDao.getPermissions(id).size());
    }
}

File: archivist/src/test/kotlin/com/zorroa/archivist/repository/BlobDaoTests.kt
package com.zorroa.archivist.repository

import com.google.common.collect.ImmutableMap
import com.zorroa.archivist.AbstractTest
import com.zorroa.archivist.domain.*
import com.zorroa.security.Groups
import org.junit.Before
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.DuplicateKeyException
import org.springframework.dao.EmptyResultDataAccessException
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class BlobDaoTests : AbstractTest() {

    @Autowired
    private lateinit var blobDao : BlobDao

    @Autowired
    private lateinit var permissionDao : PermissionDao

    lateinit var blob: Blob

    @Before
    fun init() {
        blob = blobDao.create("app", "feature", "name",
                ImmutableMap.of("foo", "bar"))
    }

    @Test
    fun testCreate() {
        assertEquals("app", blob.app)
        assertEquals("feature", blob.feature)
        assertEquals("name", blob.name)
        assertEquals("bar", (blob.data as Map<*, *>)["foo"])
    }

    @Test
    fun testUpdate() {
        assertTrue(blobDao.update(blob, ImmutableMap.of("shizzle", "mcnizzle")))
    }

    @Test
    fun testDelete() {
        assertTrue(blobDao.delete(blob))
        assertFalse(blobDao.delete(blob))
    }

    @Test(expected = DuplicateKeyException::class)
    fun testCreateDuplicate() {
        blobDao.create("app", "feature", "name",
                ImmutableMap.of("foo", "bar"))
    }

    @Test
    fun testGetAll() {
        var all = blobDao.getAll("app", "feature")
        assertEquals(1, all.size.toLong())

        blobDao.create("app", "feature", "name2",
                ImmutableMap.of("foo", "bar"))
        all = blobDao.getAll("app", "feature")
        assertEquals(2, all.size.toLong())
    }

    @Test
    fun testGetId() {
        val id = blobDao.getId("app", "feature", "name", Access.Read)
        assertEquals(id.getBlobId(), blob.getBlobId())
    }

    @Test
    fun testReplacePermissions() {
        val id = blobDao.getId("app", "feature", "name", Access.Read)
        val perm = permissionService.getPermission("zorroa::administrator")

        val _acl = Acl()
        _acl.addEntry(perm, 7)
        blobDao.setPermissions(id, SetPermissions().apply { acl=_acl; replace=true })
        assertEquals(1, blobDao.getPermissions(id).size.toLong())
    }

    @Test(expected = EmptyResultDataAccessException::class)
    fun testCheckPermissions() {
        var id = blobDao.getId("app", "feature", "name", Access.Write)
        val test = permissionDao.create(
                PermissionSpec("test", "test", description = "foo"), false)

        val _acl = Acl()
        _acl.addEntry(test, 7)

        blobDao.setPermissions(id, SetPermissions().apply { acl=_acl; replace=true })

        authenticate("user")
        id = blobDao.getId("app", "feature", "name", Access.Write)
    }

    @Test(expected = DuplicateKeyException::class)
    fun testReplacePermissionsWithDuplicate() {
        val id = blobDao.getId("app", "feature", "name", Access.Read)
        val _acl = Acl()
        val perm = permissionService.getPermission("zorroa::administrator")
        _acl.addEntry(perm, 7)
        _acl.addEntry(perm, 7)
        blobDao.setPermissions(id, SetPermissions().apply { acl=_acl; replace=true })
    }

    @Test
    fun testUpdatePermissions() {
        val id = blobDao.getId("app", "feature", "name", Access.Read)
        var _acl = Acl()
        val perm = permissionService.getPermission(Groups.ADMIN)
        _acl.addEntry(perm, 7)
        blobDao.setPermissions(id, SetPermissions().apply { acl=_acl; replace=false })
        assertEquals(1, blobDao.getPermissions(id).size)

        // add a new permission
        val p = permissionDao.get(Groups.MANAGER)
        _acl = Acl()
        _acl.addEntry(p, 7)
        blobDao.setPermissions(id, SetPermissions().apply { acl=_acl; replace=false })
        assertEquals(2, blobDao.getPermissions(id).size.toLong())

        // remove permission
        _acl = Acl()
        _acl.addEntry(p, 0)
        blobDao.setPermissions(id, SetPermissions().apply { acl=_acl; replace=false })
        assertEquals(1, blobDao.getPermissions(id).size.toLong())
    }
}
